### PR TITLE
Improve typings on `Table`, `ItemListLayout`, `KubeObjectListLayout`, `KubeApi`

### PR DESCRIPTION
**Pull request by Nokel81**
Signed-off-by: Sebastian Malton <sebastian@malton.name>

Breaking Changes:
- `KubeApi`'s constructor now requires `objectConstructor` (previously optional). This is done so that the resulting type of the `objectConstructor` (which returns `K extends KubeObject`) is "picked up" as the type parameter's value for `KubeApi`.
- Remove type parameter's default from `KubeApi`. This is required to help with the above.
- Remove type parameter's default from `IKubeObjectConstructor`. This is required to help with the first.

### CODE REVIEW

File: src/renderer/api/endpoints/resource-applier.api.ts [hunk 1]
```diff
@@ -30,7 +30,7 @@ export const resourceApplierApi = {
     "kubectl.kubernetes.io/last-applied-configuration"
   ],
 
-  async update<D extends KubeObject>(resource: object | string): Promise<D> {
+  async update<KO extends KubeObject>(resource: object | string): Promise<KO | null> {
```
Comment (jakolehm) [inline src/renderer/api/endpoints/resource-applier.api.ts]: `KO` looks a bit odd, maybe just `K`.

File: src/renderer/api/endpoints/resource-applier.api.ts [hunk 2]
```diff
@@ -48,7 +48,7 @@ export const resourceApplierApi = {
           }
         });
 
-        return items.length === 1 ? items[0] : items;
+        return items[0] as KO ?? null;
```
Comment (jakolehm) [inline src/renderer/api/endpoints/resource-applier.api.ts]: This changes behaviour but I have no idea why we have returned an item or an array based on length 🤔

File: src/renderer/api/kube-object.ts [hunk 3]
```diff
@@ -265,8 +265,8 @@ export class KubeObject implements ItemObject {
   }
 
   // use unified resource-applier api for updating all k8s objects
-  async update<T extends KubeObject>(data: Partial<T>) {
-    return resourceApplierApi.update<T>({
+  async update<KO extends KubeObject>(data: Partial<KO>) {
```
Comment (jakolehm) [inline src/renderer/api/kube-object.ts]: `KO` looks odd also here ...

File: src/renderer/api/endpoints/resource-applier.api.ts (hunk 2 above)
Comment (Nokel81) [reply src/renderer/api/endpoints/resource-applier.api.ts]: I agree that this changes behaviour, but if you look at the declared types. We never said we were returning an array. So I don't believe anyone was expecting an array.

Comment (jakolehm) [reply src/renderer/api/endpoints/resource-applier.api.ts]: Yeah, I agree.. old behaviour is a bit confusing.

File: src/renderer/components/item-object-list/item-list-layout.tsx [hunk 4]
```diff
@@ -186,16 +190,6 @@ export class ItemListLayout extends React.Component<ItemListLayoutProps> {
 
       return items;
     },
-
-    [FilterType.NAMESPACE]: items => {
```
Comment (nevalla) [inline src/renderer/components/item-object-list/item-list-layout.tsx]: This is not needed or why is this removed?

**Review by nevalla — DISMISSED** [review]

File: src/renderer/components/item-object-list/item-list-layout.tsx (hunk 4 above)
Comment (Nokel81) [reply src/renderer/components/item-object-list/item-list-layout.tsx]: Because `Items` don't necessarily have a `getNs()` command, only `KubeObjects` do. So this was moved to `KubeObjectListLayout` and passed down.

File: src/renderer/api/kube-api.ts [hunk 5]
```diff
@@ -127,15 +127,15 @@ export function ensureObjectSelfLink(api: KubeApi, object: KubeJsonApiData) {
...
+export type KubeApiWatchCallback = (data: IKubeWatchEvent<KubeJsonApiData>, error: any) => void;
 
 export type KubeApiWatchOptions = {
   namespace: string;
   callback?: KubeApiWatchCallback;
   abortController?: AbortController
 };
 
-export class KubeApi<T extends KubeObject = any> {
+export class KubeApi<T extends KubeObject> {
```
Comment (jakolehm) [inline src/renderer/api/kube-api.ts:138]: This is a breaking change, right? `KubeApi` now requires `T` because it does not have default value `any`?

Would `export class KubeApi<T extends KubeObject = KubeObject> {` work?

File: src/renderer/components/item-object-list/item-list-layout.tsx [hunk 6]
```diff
@@ -46,8 +46,10 @@ import { namespaceStore } from "../+namespaces/namespace.store";
 
 // todo: refactor, split to small re-usable components
 
-export type SearchFilter<T extends ItemObject = any> = (item: T) => string | number | (string | number)[];
-export type ItemsFilter<T extends ItemObject = any> = (items: T[]) => T[];
+export type SearchFilter<Item extends ItemObject> = (item: Item) => string | number | (string | number)[];
```
Comment (jakolehm) [inline src/renderer/components/item-object-list/item-list-layout.tsx]: In other places we are using single uppercase letter. Here we chose to use `Item`, any reason for that?

File: src/renderer/components/table/table.tsx [hunk 7]
```diff
@@ -74,16 +74,16 @@ export const orderByUrlParam = createPageParam({
 });
 
 @observer
-export class Table extends React.Component<TableProps> {
-  static defaultProps: TableProps = {
+export class Table<Item> extends React.Component<TableProps<Item>> {
```
Comment (jakolehm) [inline src/renderer/components/table/table.tsx:77]: Table improvements are also breaking changes.

File: src/renderer/kube-object.store.ts [hunk 8]
```diff
@@ -31,16 +31,16 @@ import { IKubeApiQueryParams, KubeApi, parseKubeApi } from "./api/kube-api";
...
+  api?: KubeApi<K>;
   reqInit?: RequestInit;
 }
 
-export abstract class KubeObjectStore<T extends KubeObject = any> extends ItemStore<T> {
+export abstract class KubeObjectStore<K extends KubeObject> extends ItemStore<K> {
   static defaultContext = observable.box<ClusterContext>(); // TODO: support multiple cluster contexts
 
-  abstract api: KubeApi<T>;
+  abstract api: KubeApi<K>;
```
Comment (jakolehm) [inline src/renderer/kube-object.store.ts]: All changes in this file after this line are just because `T` is now `K`?

File: src/renderer/kube-object.store.ts [hunk 9]
```diff
@@ -406,7 +407,7 @@ export abstract class KubeObjectStore<T extends KubeObject = any> extends ItemSt
       switch (type) {
         case "ADDED":
         case "MODIFIED":
-          const newItem = new api.objectConstructor(object);
+          const newItem = new api.objectConstructor(object) as K;
```
Comment (jakolehm) [inline src/renderer/kube-object.store.ts]: Do we need `as K` here?

File: src/renderer/api/kube-api.ts (hunk 5 above)
Comment (Nokel81) [reply src/renderer/api/kube-api.ts:138]: Yes that would work. I can change it back.

File: src/renderer/components/item-object-list/item-list-layout.tsx (hunk 6 above)
Comment (Nokel81) [reply src/renderer/components/item-object-list/item-list-layout.tsx]: Not really, I guess I could change these to `I` instead.

File: src/renderer/api/kube-api.ts (hunk 5 above)
Comment (Nokel81) [reply src/renderer/api/kube-api.ts:138]: It is a breaking change yes, though only if you take it as an argument, you don't need to specify it when doing `new KubeApi(...)`.

Furthermore, from a TS perspective it is better not to have this as a default parameter because then the passed in `objectConstructor` doesn't get to "set" this type parameter.

File: src/renderer/components/table/table.tsx (hunk 7 above)
Comment (Nokel81) [reply src/renderer/components/table/table.tsx:77]: Are they? I don't think so, it isn't required to use the new `renderRow` function (just helps with typings) and it is not required to specify the type arguments on use. (if you don't specify any, currently it is all of them or none of them).

File: src/renderer/kube-object.store.ts (hunk 8 above)
Comment (Nokel81) [reply src/renderer/kube-object.store.ts]: Mostly

File: src/renderer/kube-object.store.ts (hunk 9 above)
Comment (Nokel81) [reply src/renderer/kube-object.store.ts]: Not if I removed the default type argument from `IKubeObjectConstructor`. Which imo is the better solution.

File: src/renderer/api/kube-api.ts (hunk 5 above)
Comment (jakolehm) [reply src/renderer/api/kube-api.ts:138]: Please describe this in the PR description (maybe PR title also needs an update?). I also think we need to add `area/extension` label because this affects extension API.

File: src/renderer/components/table/table.tsx (hunk 7 above)
Comment (jakolehm) [reply src/renderer/components/table/table.tsx:77]: Ok 👍  I somehow misinterpreted the diff here.

File: src/renderer/components/table/table.tsx [hunk 10]
```diff
@@ -171,9 +171,20 @@ export class Table extends React.Component<TableProps> {
     });
   }
 
-  renderRows() {
-    const { sortable, noItems, children, virtual, customRowHeights, rowLineHeight, rowPadding, items, getTableRow, selectedItemId, className } = this.props;
+  getContent() {
```
Comment (jakolehm) [inline src/renderer/components/table/table.tsx]: Should this be private/protected if it's just used within `renderRows`?

File: src/renderer/kube-object.store.ts (hunk 8 above)
Comment (jakolehm) [reply src/renderer/kube-object.store.ts]: These kind of changes makes reviewing hard. What's wrong with `T ` and why we wanted to change existing practice (single letter vs word) in this PR 🤔

File: src/renderer/components/table/table.tsx (hunk 10 above)
Comment (Nokel81) [reply src/renderer/components/table/table.tsx]: Sure... for react components I haven't been able to figure out why `private` or `protected` are good ideas for the instance methods. Because it is not like we call the constructor ourselves.

File: src/renderer/kube-object.store.ts (hunk 8 above)
Comment (Nokel81) [reply src/renderer/kube-object.store.ts]: I can revert them, I change it because `K` is closer to `KubeObject` which is what it represents.
